proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,356 @@
+  - If this is not an initialization:
+    - Verify the program account
+    - Check that the status stored in the program account is retracted,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption of the new length, otherwise throw `InsufficientFunds`
+  - If the requested new size is zero:
+    - Set the length of the program account to 0 (removing the header too)
+    - Assign ownership of the program account to the system program
+  - If the requested new size is greater than zero:
+    - Set the length of the program account to the requested new size plus
+    the header size
+    - In case that this is an initialization, also initialize the header:
+      - Set the slot to zero, **not** the current slot
+      - Set the authority address (from the instruction account at index 1)
+      - Set the status to retracted
+
+#### Deploy
+
+- Instruction accounts:
+  - `[writable]` The program account to deploy.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `3u32`
+- Behavior:
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account
+  - Check that the slot stored in the program account is not the current
+  (deployment cooldown), otherwise throw `InvalidArgument`
+  - Note: The cooldown enforces that each pair of an address and a slot can
+  uniquely identify a deployment of a program, which simplifies caching logic.
+  - Check that the status stored in the program account is retracted
+    otherwise throw `InvalidArgument`
+  - Check that the executable file stored in the program account passes
+  executable verification
+  - Change the slot in the program account to the current slot
+  - Change the status stored in the program account to deployed
+  - Set the `is_executable` flag to `true`
+
+#### Retract
+
+- Instruction accounts:
+  - `[writable]` The program account to retract.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `4u32`
+- Behavior:
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account
+  - Check that the slot stored in the program account is not the current
+  (deployment cooldown), otherwise throw `InvalidArgument`
+  - Check that the status stored in the program account is deployed,
+    otherwise throw `InvalidArgument`
+  - Note: The slot is **not** set to the current slot to allow a
+  retract-modify-redeploy-sequence within the same slot or even within the
+  same transaction.
+  - Change the status stored in the program account to retracted
+  - Set the `is_executable` flag to `false`
+
+#### TransferAuthority
+
+- Instruction accounts:
+  - `[writable]` The program account to change the authority of.
+  - `[signer]` The current authority of the program.
+  - `[signer]` The new authority of the program.
```

Comment:
I assume the reason to remove the _unchecked_ version of this instruction is due to concerns of losing the authority. Assuming this is the case, could we have an alternative 2-step process (TransferAuthority + AcceptTransferAuthority)? 

With the ideas around removing nonces, requiring both signatures is going to further increase developer frustrations around program upgrades.

Reply:
How would you implement that? It seems it would require a transitory state (`InAuthorityTransfer`) which would require two authority fields in the header in case the new one does not accept and you need to revert.